File: github_cpp/14/36.cpp








#include "sort.h"
#include <iostream>
#include "string.h"
#include <cmath>



 
int getMax(int A[], int size)
{
    int mx = A[0];
    for (int i = 1; i < size; i++)
        if (A[i] > mx)
            mx = A[i];
    return mx;
}
 


void countSort(int A[], int size, int exp)
{
    int output[size]; 
    int i, count[10] = {0};
 
    
    for (i = 0; i < size; i++)
        count[ (A[i]/exp)%10 ]++;
 
    
    
    for (i = 1; i < 10; i++)
        count[i] += count[i - 1];
 
    
    for (i = size - 1; i >= 0; i--)
    {
        output[count[ (A[i]/exp)%10 ] - 1] = A[i];
        count[ (A[i]/exp)%10 ]--;
    }
 
    
    
    for (i = 0; i < size; i++)
        A[i] = output[i];
}
 

void
RadixSort::sort(int A[], int size)
{
    
    int m = getMax(A, size);
 
    
    
    for (int exp = 1; m/exp > 0; exp *= 10)
        countSort(A, size, exp);
}

File: github_java_sort/6/36.java
package com.jnu.sortDemo;

import java.util.ArrayList;
import java.util.List;


public class RadixSort {

    public static void radixSort(int[] arr){
        int radix = getRadix(arr);
        sort(arr, radix);
    }

    private static void sort(int[] arr, int radix){
        
        List<List<Integer>> list = new ArrayList<List<Integer>>();
        for(int i = 0; i < 10; i ++){
            List<Integer> item = new ArrayList<>();
            list.add(item);
        }

        
        for(int i = 0; i < radix; i ++){
            
            for(int j = 0; j < arr.length; j ++){
                int index = arr[j] % (int)Math.pow(10, i+1) / (int)Math.pow(10, i);
                list.get(index).add(arr[j]);
            }
            
            int count = 0;
            for(int k = 0; k < 10; k ++){
                if(list.get(k).size() > 0){
                    for(int temp : list.get(k)){
                        arr[count] = temp;
                        count ++;
                    }
                    
                    list.get(k).clear();
                }
            }
        }
    }

    
    private static int getRadix(int[] arr) {
        int max = getMax(arr);
        int time = 0;
        while ( max > 0){
            max /= 10;
            time ++;
        }
        return time;
    }

    private static int getMax(int[] arr){
        int flag = 0;
        for(int i = 0; i < arr.length; i ++){
            flag = i;
            for(int j = i + 1; j < arr.length; j ++){
                if(arr[j] > arr[flag]){
                    flag = j;
                }
            }
            if(flag != i){
                break;
            }
        }
        return arr[flag];
    }

}
